Add ability to migrate with promises.

KenPowers opened the pull request:
These changes enable the use of promises in migrations. For example, the following:

```js
var async = require('async');

exports.up = function (db, callback) {
  async.series([
    db.createTable.bind(db, 'pets', {
      id: { type: 'int', primaryKey: true },
      name: 'string'
    }),
    db.createTable.bind(db, 'owners', {
      id: { type: 'int', primaryKey: true },
      name: 'string'
    });
  ], callback);
};

exports.down = function (db, callback) {
  async.series([
    db.dropTable.bind(db, 'pets'),
    db.dropTable.bind(db, 'owners')
  ], callback);
};
```

Can be turned into the following:

```js
exports.up = function (db, callback) {
  return db.createTableAsync('pets', {
    id: {type: 'int', primaryKey: true },
    name: 'string'
  }).then(function () {
    return db.createTableAsync('owners', {
      id: { type: 'int', primaryKey: true },
      name: 'string'
    });
  });
};

exports.down = function (db, callback) {
  return db.dropTableAsync('pets').then(function () {
    return db.dropTableAsync('owners');
  });
};
```

Imagine chaining like crazy:

```js
exports.down = function (db, callback) {
  return db.dropTableAsync('pets').then(function () {
    return db.dropTableAsync('owners');
  }).then(function () {
    return db.dropTableAsync('third');
  }).then(function () {
    return db.dropTableAsync('fourth');
  }).then(function () {
    return db.dropTableAsync('fifth');
  });
};
```

The real benefit is how much control-flow stuff we gain from Bluebird without having to include any additional libraries (everything is built in to the promises): https://github.com/petkaantonov/bluebird/blob/master/API.md

Just for reference, this will improve even more with ES6:

```js
exports.down = function (db, callback) {
  return db.dropTable('pets').then(() => db.dropTable('owners'));
};
```

Something similar is currently possible with LoDash:

```js
exports.down = function (db, callback) {
  return db.dropTable('pets').then(_.bindKey(db, 'dropTable', 'owners'));
};
```

The Bluebird `Promise` constructor is exported on `index`:

```js
var Promise = require('db-migrate').Promise;
```

What still needs to be done:

* Documentation
* Tests
* Update migration template to get a reference to `Promise` (optional)
* Feedback

wzrdtales commented:
@KenPowers I'm going to take look at this over the next.

wzrdtales commented:
@KenPowers Could you please **rebase** against the master, having promises would be great, as they make handling async actions much easier.

Also promises are a good and modern replacement for async, so I would like looking forward to get them into db-migrate. Also they're faster then async and have a smaller footprint, so I think promises should be definetly available within db-migrate. 

Yet there are missing tests, so that has definetly to be done to merge this. Also I wouldn't update the template, instead there should be added an additional one.

wzrdtales commented:
closed and reopened to initiate travis

KenPowers commented:
I started the rebase and I think it's probably just easier to start again for the following reasons:

1. The project has advanced a lot since this pull request was opened.
1. My changes were simple.
1. I split my commit off on to its own branch, accepted the latest upstream change on to master, rebased my branch on master (which I'm currently in the middle of a hairy merge), but even after all of that this pull request becomes defunct because my changes are on a different branch (yes yes, I should have made changes off of master).

wzrdtales commented:
You're doing it probably wrong...

wzrdtales commented:
The rebase is really simple to handle, I did just the following:

git fetch https://github.com/KenPowers/node-db-migrate.git
git checkout FETCH_HEAD
git rebase origin/master

Conflict happens on package.json
git mergetool

Take everything from the actual master and add bluebird to it.
git rebase --continue

You have completely finished the rebase.

wzrdtales commented:
So something you're doing wrong, maybe you have rebased the wrong direction? you're in the master and executed
git rebase yourchanges

This is wrong, you need to rebase against the master from your branch with your changes.

So you probably do the following on the branch with your changes.
git pull --rebase git@github.com:kunklejr/node-db-migrate.git master

KenPowers commented:
What I did was:

```sh
# Get my changes on their own branch.
git branch my-changes
# Get my changes off of master.
git reset --hard HEAD~1
# Go to my new branch.
git checkout my-changes
# Rebase on master.
git rebase master
# Handle conflicts.
git mergetool
```

And yes, I realize the merge was simple. What had happened was my version of `npm` was alphabetizing dependencies and you had different dependency versions. Admittedly I didn't even try to merge at that point.

What version of `npm` are you using, by the way? My `npm` has been alphabetizing for a long time now.

wzrdtales commented:
I use the latest npm.

Am 09.02.2015 um 13:59 schrieb Kenneth Powers:
>
> What I did was:
>
> # Get my changes on their own branch.
> git branch my-changes
> # Get my changes off of master.
> git reset --hard HEAD~1
> # Go to my new branch.
> git checkout my-changes
> # Rebase on master.
> git rebase master
> # Handle conflicts.
> git mergetool
>
> And yes, I realize the merge was simple. What had happened was my
> version of |npm| was alphabetizing dependencies and you had different
> dependency versions. Admittedly I didn't even try to merge at that point.
>
> What version of |npm| are you using, by the way? My |npm| has been
> alphabetizing for a long time now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kunklejr/node-db-migrate/pull/194#issuecomment-73504376>.
>

wzrdtales commented:
Planned for v0.10.0 for now.

wzrdtales commented:
closed in favour of #239

rosendi commented:
@wzrdtales Does it support async/await feature now?

wzrdtales commented:
@rosendi In 0.10.x(-beta.x) promises are included and the default.

wzrdtales commented:
@rosendi Had not that much time answer that day.
Promises are included thus async await could be used in combination with babel. Promises with a recent node version (4 or 5) is quite straightforward already.

So this example:

```javascript
exports.down = function (db) {
  return db.dropTableAsync('pets')
    .then(() => db.dropTableAsync('owners'))
    .then(() => db.dropTableAsync('third'))
    .then(() => db.dropTableAsync('fourth'))
    .then(() => db.dropTableAsync('fifth'));
}
```

is already possible today. Using babel with db-migrate is something that is on the list, but not yet worked on.

rosendi commented:
@wzrdtales Sorry, I forgot to reply. I used 0.9.x and that was the point. Thank you for your time!
